### PR TITLE
Speakers Promo Codes

### PR DESCRIPTION
**Pull request by romanetar**
ref https://tipit.avaza.com/project/view#!tab=task-pane&groupby=MyTaskDueDate&view=vertical&task=3270905&stafilter=NotComplete&fileview=grid

Signed-off-by: romanetar <roman_ag@hotmail.com>

### CODE REVIEW

File: app/Models/Foundation/Summit/Registration/PromoCodes/Traits/SpeakersPromoCodeTrait.php [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    {
+        return count($this->owners) > 0;
+    }
+
+    /**
+     * @return AssignedPromoCodeSpeaker[]
+     */
+    public function getOwners(): array
+    {
+        return $this->owners->toArray();
```
Comment (smarcet) [inline app/Models/Foundation/Summit/Registration/PromoCodes/Traits/SpeakersPromoCodeTrait.php:56]: his is not a good approach bc get all collection in memory

File: app/Models/Foundation/Summit/Registration/PromoCodes/Traits/SpeakersPromoCodeTrait.php [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+     * @param string $type
+     */
+    public function setType($type)
+    {
+        $this->type = $type;
+    }
+
+    public function hasOwners(): bool
+    {
+        return count($this->owners) > 0;
```
Comment (smarcet) [inline app/Models/Foundation/Summit/Registration/PromoCodes/Traits/SpeakersPromoCodeTrait.php:48]: count should be invoked on method ->count()

File: app/Models/Foundation/Summit/Registration/PromoCodes/Traits/SpeakersPromoCodeTrait.php (hunk 1 above)
Comment (smarcet) [reply app/Models/Foundation/Summit/Registration/PromoCodes/Traits/SpeakersPromoCodeTrait.php:56]: owners should be declared extra lazy
check https://www.doctrine-project.org/projects/doctrine-orm/en/2.15/tutorials/extra-lazy-associations.html

File: app/Services/Model/Imp/SpeakerService.php [hunk 3]
```diff
@@ -1325,9 +1247,11 @@ function($summit, $flow_event, $speaker_id, $test_email_recipient, $speaker_anno
                         $speaker_id,
                         $test_email_recipient,
                         $speaker_announcement_email_config,
-                        $filter
+                        $filter,
+                        $payload
                     ) {
                         $email_strategy = new SpeakerActionsEmailStrategy($summit, $flow_event);
+                        $promo_code_strategy_factory = new PromoCodeStrategyFactory($this->tx_service, $this->code_generator);
```
Comment (smarcet) [inline app/Services/Model/Imp/SpeakerService.php:1254]: factory should be injected at IOC level

File: app/Services/Model/Imp/SummitOrderService.php [hunk 4]
```diff
@@ -2331,7 +2331,8 @@ private function createTicketsForOrder
                         throw new EntityNotFoundException(sprintf("Promo code %s not found.", $promo_code));
                     }
                     Log::debug(sprintf("SummitOrderService::createTicketsForOrder applying promo code %s", $pc->getCode()));
-                    $pc->addUsage(1);
+                    $owner_email = !is_null($attendee) ? $attendee->getEmail() : $order->getOwnerEmail();
+                    $pc->addUsage(1, $owner_email);
```
Comment (smarcet) [inline app/Services/Model/Imp/SummitOrderService.php:2335]: lets invert the signature of addUsage
addUsage($owner_email, $qty = 1) so we could invoke here
like
$pc->addUsage($owner_email)

File: database/migrations/model/Version20230428191955.php [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+        $builder = new Builder($schema);
+        if (!$builder->hasTable("AssignedPromoCodeSpeaker")) {
+            $builder->create("AssignedPromoCodeSpeaker", function (Table $table) {
+                $table->integer("ID", true, false);
+                $table->primary("ID");
+
+                $table->timestamp("RedeemedAt")->setNotnull(false);
+                $table->timestamp("SentAt")->setNotnull(false);
+
+                $table->integer("RegistrationPromoCodeID")->setNotnull(false)->setDefault('NULL');
```
Comment (smarcet) [inline database/migrations/model/Version20230428191955.php:36]: this should be only one id
pointing to the basa table

Comment (smarcet) [inline database/migrations/model/Version20230428191955.php:36]: this should be pointing to  base table 
https://github.com/OpenStackweb/summit-api/blob/main/app/Models/Foundation/Summit/Registration/PromoCodes/SummitRegistrationPromoCode.php

File: database/migrations/model/Version20230428191955.php [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+                $table->foreign("SpeakersSummitRegistrationPromoCode",
+                    "RegistrationPromoCodeID",
+                    "ID",
+                    ["onDelete" => "CASCADE"],
+                    "FK_AssignedPromoCodeSpeaker_RegistrationPromoCode");
+
+                $table->integer("RegistrationDiscountCodeID")->setNotnull(false)->setDefault('NULL');
```
Comment (smarcet) [inline database/migrations/model/Version20230428191955.php:43]: this is not needed see comment above

File: app/Models/Foundation/Summit/Registration/PromoCodes/AssignedPromoCodeSpeaker.php [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+     * @var SpeakersSummitRegistrationPromoCode
+     */
+    protected $registration_promo_code;
+
+    /**
+     * @ORM\ManyToOne(targetEntity="SpeakersRegistrationDiscountCode")
+     * @ORM\JoinColumn(name="RegistrationDiscountCodeID", referencedColumnName="ID")
+     * @var SpeakersRegistrationDiscountCode
+     */
+    protected $registration_discount_code;
```
Comment (smarcet) [inline app/Models/Foundation/Summit/Registration/PromoCodes/AssignedPromoCodeSpeaker.php:43]: not needed

File: app/Models/Foundation/Summit/Registration/PromoCodes/AssignedPromoCodeSpeaker.php [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+    {
+        return $this->registration_promo_code;
+    }
+
+    /**
+     * @param SpeakersSummitRegistrationPromoCode $registration_promo_code
+     */
+    public function setRegistrationPromoCode(SpeakersSummitRegistrationPromoCode $registration_promo_code): void
+    {
+        $this->registration_promo_code = $registration_promo_code;
```
Comment (smarcet) [inline app/Models/Foundation/Summit/Registration/PromoCodes/AssignedPromoCodeSpeaker.php:107]: u should check here if its an instance of the expected relationship ( instance of SpeakersSummitRegistrationPromoCode or SpeakersRegistrationDiscountCode

**Review by smarcet — CHANGES_REQUESTED** [review]
@romanetar please review changes

File: app/Models/Foundation/Summit/Registration/PromoCodes/Traits/SpeakersPromoCodeTrait.php (hunk 2 above)
Comment (romanetar) [reply app/Models/Foundation/Summit/Registration/PromoCodes/Traits/SpeakersPromoCodeTrait.php:48]: done

File: app/Models/Foundation/Summit/Registration/PromoCodes/Traits/SpeakersPromoCodeTrait.php [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+    /**
+     * @param int $usage
+     * @param string $owner_email
+     * @throws \Exception
+     */
+    public function addUsage(int $usage, string $owner_email)
+    {
+        try {
+            $existing_owner = $this->owners->filter(function ($e) use($owner_email){
+                $member = $e->getSpeaker()->getMember();
```
[the remark this answers is not in the document]
Comment (romanetar) [reply app/Models/Foundation/Summit/Registration/PromoCodes/Traits/SpeakersPromoCodeTrait.php:80]: done

File: app/Services/Model/Imp/SummitOrderService.php (hunk 4 above)
Comment (romanetar) [reply app/Services/Model/Imp/SummitOrderService.php:2335]: done

File: database/migrations/model/Version20230428191955.php (hunk 6 above)
Comment (romanetar) [reply database/migrations/model/Version20230428191955.php:43]: done

File: app/Models/Foundation/Summit/Registration/PromoCodes/AssignedPromoCodeSpeaker.php (hunk 7 above)
Comment (romanetar) [reply app/Models/Foundation/Summit/Registration/PromoCodes/AssignedPromoCodeSpeaker.php:43]: done

File: app/Models/Foundation/Summit/Registration/PromoCodes/AssignedPromoCodeSpeaker.php (hunk 8 above)
Comment (romanetar) [reply app/Models/Foundation/Summit/Registration/PromoCodes/AssignedPromoCodeSpeaker.php:107]: done

File: database/migrations/model/Version20230428191955.php (hunk 5 above)
Comment (romanetar) [reply database/migrations/model/Version20230428191955.php:36]: done

File: app/Models/Foundation/Summit/Registration/PromoCodes/Traits/SpeakersPromoCodeTrait.php (hunk 1 above)
Comment (romanetar) [reply app/Models/Foundation/Summit/Registration/PromoCodes/Traits/SpeakersPromoCodeTrait.php:56]: done

File: app/Services/Model/Imp/SpeakerService.php (hunk 3 above)
Comment (romanetar) [reply app/Services/Model/Imp/SpeakerService.php:1254]: done

**Comment by romanetar** [comment]
@smarcet all the review changes have been addressed

File: app/Services/Model/Imp/SummitOrderService.php [hunk 10]
```diff
@@ -594,8 +594,8 @@ public function run(array $formerState): array
 
                 Log::debug(sprintf("adding %s usage to promo code %s", $qty, $promo_code->getId()));
 
-                $this->lock_service->lock('promocode.' . $promo_code->getId() . '.usage.lock', function () use ($promo_code, $qty) {
-                    $promo_code->addUsage($qty);
+                $this->lock_service->lock('promocode.' . $promo_code->getId() . '.usage.lock', function () use ($promo_code, $qty, $owner_email) {
```
Comment (smarcet) [inline app/Services/Model/Imp/SummitOrderService.php:597]: we missed to update the compensatory transaction here 
https://github.com/OpenStackweb/summit-api/blob/f084b54f065992303c49f53b5ba23d022b4cacc0/app/Services/Model/Imp/SummitOrderService.php#L620
we need to update the signature of removeUsage here
https://github.com/OpenStackweb/summit-api/blob/f084b54f065992303c49f53b5ba23d022b4cacc0/app/Models/Foundation/Summit/Registration/PromoCodes/SummitRegistrationPromoCode.php#L363
to accept an optional owner_email address and override the proper logic on the speakers promo code trait

Comment (smarcet) [reply app/Services/Model/Imp/SummitOrderService.php:597]: also we need to review all the uses of removeUsage ( whole project)

**Review by smarcet — CHANGES_REQUESTED** [review]
@romanetar please review comments

File: app/Services/Model/Strategies/PromoCodes/AutomaticMultiSpeakerPromoCodeStrategy.php [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+     * @return SummitRegistrationPromoCode|null
+     * @throws EntityNotFoundException
+     * @throws ValidationException|\Exception
+     */
+    public function getPromoCode(PresentationSpeaker $speaker): ?SummitRegistrationPromoCode {
+
+        return $this->tx_service->transaction(function () use ($speaker) {
+            $code = null;
+            do {
+                $code = $this->code_generator->generate($this->summit, 6);
```
Comment (smarcet) [inline app/Services/Model/Strategies/PromoCodes/AutomaticMultiSpeakerPromoCodeStrategy.php:86]: this magic number should be a constant of the generator

File: app/Services/Model/Strategies/PromoCodes/PromoCodeGenerator.php [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+        return Rand::getString($length, self::VsChar);
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function generate(Summit $summit, int $length): string
+    {
+        do {
+            $key = sprintf("%s_%s", $summit->getName(), $this->generateValue($length));
```
Comment (smarcet) [inline app/Services/Model/Strategies/PromoCodes/PromoCodeGenerator.php:55]: length should be a constant of the generator and passs it on the constructor IOC

Comment (smarcet) [inline app/Services/Model/Strategies/PromoCodes/PromoCodeGenerator.php:55]: we need to ensure strupper here

Comment (smarcet) [inline app/Services/Model/Strategies/PromoCodes/PromoCodeGenerator.php:55]: we should use https://github.com/OpenStackweb/summit-api/blob/a6988565fe0c7c6da9458009f2a0a08ecc42188e/app/Models/Foundation/Summit/Summit.php#LL2201C21-L2201C46

Comment (romanetar) [reply app/Services/Model/Strategies/PromoCodes/PromoCodeGenerator.php:55]: done

Comment (romanetar) [reply app/Services/Model/Strategies/PromoCodes/PromoCodeGenerator.php:55]: done

File: app/Services/Model/Strategies/PromoCodes/AutomaticMultiSpeakerPromoCodeStrategy.php (hunk 11 above)
Comment (romanetar) [reply app/Services/Model/Strategies/PromoCodes/AutomaticMultiSpeakerPromoCodeStrategy.php]: done

File: app/Services/Model/Strategies/PromoCodes/PromoCodeGenerator.php (hunk 12 above)
Comment (romanetar) [reply app/Services/Model/Strategies/PromoCodes/PromoCodeGenerator.php]: done

File: app/Models/Foundation/Summit/Registration/PromoCodes/SummitRegistrationPromoCode.php [hunk 13]
```diff
@@ -355,9 +358,10 @@ public function addUsage(int $usage)
 
     /**
      * @param int $to_restore
+     * @param string|null $owner_email
      * @throws ValidationException
      */
-    public function removeUsage(int $to_restore)
+    public function removeUsage(int $to_restore, string $owner_email = null)
```
Comment (smarcet) [inline app/Models/Foundation/Summit/Registration/PromoCodes/SummitRegistrationPromoCode.php:364]: this should be ?string

**Pull request merged**